design/src/rfcs/rfc0026_custom_validation_exceptions.md

```diff
@@ -0,0 +1,671 @@
+RFC: Custom validation exceptions
+=================================
+
+> Status: RFC
+>
+> Applies to: server
+
+This RFC describes a feature whereby server SDK application owners will be able
+to return custom error messages when validation of [constraint
+traits][constraint-traits] fails for an operation's input.
+
+[constraint-traits]: https://awslabs.github.io/smithy/2.0/spec/constraint-traits.html
+
+Terminology
+-----------
+
+- **Shape closure**: the set of shapes a shape can "reach", including itself.
+- **Transitively constrained shape**: a shape whose closure includes: 
+    1. a shape with a [constraint trait][constraint-traits] attached,
+    2. a (member) shape with a [`required` trait] attached,
+    3. an [`enum` shape]; or
+    4. an [`intEnum` shape].
+- A **directly constrained shape** is any of these:
+    1. a shape with a [constraint trait][constraint-traits] attached,
+    2. a (member) shape with a [`required` trait] attached,
+    3. an [`enum` shape],
+    4. an [`intEnum` shape]; or
+    5. a [`structure` shape] with at least one `required` member shape.
+- **Constrained type**: the Rust type a constrained shape gets rendered as. For
+  shapes that are not `structure`, `union`, `enum`, or `intEnum` shapes, these
+  are wrapper [newtype]s.
+- **`ConstraintViolation`**: the Rust type a constrained type's [`TryFrom`]
+  constructor exposes to user code when a value violates the direct modeled
+  constraints.
+- **`ConstraintViolationException`**: the Rust type a constrained type's
+  _internal_ framework-exclusive constructor returns when a value violates the
+  modeled constraints.
+- **Constrained operation**: an operation whose input shape is transitively
```

Comment:
Because the framework doesn't care about constraints on output or error shapes insofar as they are something the user constructs and hence the constrained types take care of honoring the constraints _in their application code_. In this RFC we're just interested in the constraints of shapes that lie in operation inputs, because those are checked _in framework code_ upon request deserialization.